capsules/src/bmp280.rs

```diff
@@ -0,0 +1,469 @@
+    /// xlsb: [7:4]
+    PRESS_XLSB = 0xf9,
+    TEMP_MSB = 0xfa,
+    TEMP_LSB = 0xfb,
+    /// xlsb: [7:4]
+    TEMP_XLSB = 0xfc,
+}
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+struct CalibrationData {
+    dig_t1: u16,
+    dig_t2: i16,
+    dig_t3: i16,
+    // TODO: pressure calibration
+}
+
+/// CAUTION: calibration data puts least significant byte in the lowest address,
+/// readouts do the opposite.
+#[inline]
+fn twobyte(lsb: u8, msb: u8) -> u16 {
+    ((msb as u16) << 8) + lsb as u16
+}
+
+impl CalibrationData {
+    fn new(i2c_raw: &[u8]) -> Self {
+        CalibrationData {
+            dig_t1: twobyte(i2c_raw[0], i2c_raw[1]) as u16,
+            dig_t2: twobyte(i2c_raw[2], i2c_raw[3]) as i16,
+            dig_t3: twobyte(i2c_raw[4], i2c_raw[5]) as i16,
+        }
+    }
+
+    fn temp_from_raw(&self, raw_temp: u32) -> i32 {
+        let temp = raw_temp as i32; // guaranteed to succeed because raw temp has only 20 significant bits maximum.
+        let dig_t1 = self.dig_t1 as i32; // same, 16-bits
+        let dig_t2 = self.dig_t2 as i32; // same, 16-bits
+        let dig_t3 = self.dig_t3 as i32; // same, 16-bits
+                                         // From the datasheet
+        let var1 = (((temp >> 3) - (dig_t1 << 1)) * dig_t2) >> 11;
+        let a = (temp >> 4) - dig_t1;
+        let var2 = (((a * a) >> 12) * dig_t3) >> 14;
+        let t_fine = var1 + var2;
+        ((t_fine * 5) + 128) >> 8
+    }
+}
+
+/// Internal state.
+/// Each state can lead to the next on in order of appearance.
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum State {
+    Uninitialized,
+    InitId,
+    /// It's not guaranteed that the MCU reset is the same as device power-on,
+    /// so an explicit reset is necessary.
+    InitResetting,
+    InitWaitingReady,
+    InitReadingCalibration,
+
+    Idle(CalibrationData),
+
+    // States related to sample readout
+    /// One-shot mode request sent
+    Configuring(CalibrationData),
+    /// Sampling takes milliseconds, so spend most of that time sleeping.
+    WaitingForAlarm(CalibrationData),
+    /// Polling for the readout to become ready.
+    Waiting(CalibrationData),
+    /// Waiting for readout to return.
+    /// This state can also lead back to Idle.
+    Reading(CalibrationData),
+
+    /// Reset cannot be attempted.
+    /// This is because reset failed before, or because the ID is mismatched.
+    IrrecoverableError,
+    /// Irrecoverable. Currently only when init fails.
+    Error,
+}
+
+/// Complies with the reading and writing protocol used by the sensor.
+struct I2cWrapper<'a> {
+    i2c: &'a dyn i2c::I2CDevice,
+}
+
+impl<'a> I2cWrapper<'a> {
+    fn write<const COUNT: usize>(
+        &self,
+        buffer: &'static mut [u8],
+        addr: Register,
+        data: [u8; COUNT],
+    ) -> Result<(), (i2c::Error, &'static mut [u8])> {
+        buffer[0] = addr as u8;
+        for (i, d) in data.iter().enumerate() {
+            buffer[i + 1] = *d;
+        }
```

Comment:
Perhaps you can use the more elegant [`copy_from_slice`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) here, with appropriate subslicing and length checks.

Reply:
Done.
I decided to skip length checks. At the moment I'm not ready to have indexing errors propagate all the way up and push the driver into a buggy state. This is the same situation as in the thread about buggy drivers above, and handling that correctly needed a serious update.

On the bright side, all the i2c read/write code paths are utilized before a measurement can be made, so they are flawless index-wise at the moment.